internal/venice-common/src/main/java/com/linkedin/venice/writer/CompositeVeniceWriter.java

```diff
@@ -0,0 +1,182 @@
+package com.linkedin.venice.writer;
+
+import static com.linkedin.venice.writer.VeniceWriter.APP_DEFAULT_LOGICAL_TS;
+import static com.linkedin.venice.writer.VeniceWriter.DEFAULT_LEADER_METADATA_WRAPPER;
+
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.pubsub.api.PubSubProduceResult;
+import com.linkedin.venice.pubsub.api.PubSubProducerCallback;
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+
+
+/**
+ * The composite writer contains a main writer and multiple child writers. The main writer will only perform the write
+ * once all of its child writers are complete.
+ * TODO The child writers are view writers. Ideally to avoid code duplication we should be using an array of
+ * VeniceViewWriter here. However, the current implementation of VeniceViewWriter involves PCS which is something
+ * specific to the ingestion path that we don't want to leak into venice-common.
+ */
+public class CompositeVeniceWriter<K, V, U> extends AbstractVeniceWriter<K, V, U> {
+  private final VeniceWriter<K, V, U> mainWriter;
+  private final VeniceWriter<K, V, U>[] childWriters;
+  private final PubSubProducerCallback childCallback;
+
+  private CompletableFuture<PubSubProduceResult> lastWriteFuture = CompletableFuture.completedFuture(null);
+
+  public CompositeVeniceWriter(
+      String topicName,
+      VeniceWriter<K, V, U> mainWriter,
+      VeniceWriter<K, V, U>[] childWriters,
+      PubSubProducerCallback childCallback) {
+    super(topicName);
+    this.mainWriter = mainWriter;
+    this.childWriters = childWriters;
+    this.childCallback = childCallback;
+  }
+
+  @Override
+  public void close(boolean gracefulClose) throws IOException {
```

Comment:
Normally, the class should only cleanup the instances created on its own, and since both `mainWriter` and `childWriters` are created outside, in theory, these instances should be closed elsewhere. 
If we truly want to create and close these writers here, we can pass two suppliers and one for main writer and the other one for child writers and we can trigger the creation in the constructor.